### PR TITLE
add global_setup for C++ includes

### PR DESCRIPTION
**Pull request by robieta**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #51123 Instruction count benchmark (7): Use  when collecting callgrind.
* #51122 Intermezzo: Add Timer support for callgrind repeats
* #51121 Intermezzo: quality of life tweaks to Timer. (Retain callgrind.out, change collect_baseline default, add __mul__ for FunctionCounts)
* #50610 Instruction count benchmark (6): Add frontend to drive microbenchmarks, and start README
* #50609 Instruction count benchmark (5): Add benchmark runner.
* #50608 Instruction count benchmark (4): Extend the benchmarks in both coverage and complexity.
* #50607 Instruction count benchmark (3): Enable TorchScript benchmarks
* #50606 Instruction count benchmark (2): Initial benchmark definitions.
* #50605 Instruction count benchmark (1): Types and interfaces
* **#50604 add global_setup for C++ includes**
* #50603 add utility to back test Timer

In Python, it's fine to import inside a function. E.g.
```
def my_fn():
  import foo
  return foo.bar()
```

whereas in C++ the equivalent:

```
auto my_fn() {
  #include <foo>
  return foo::bar();
}
```

is at best scandalous and at more often just illegal. However it is sometimes necessary to add top level setup like `#include`s to snippets. This PR adds a C++ only `global_setup` argument to Timer.

### CODE REVIEW

File: torch/utils/benchmark/utils/timeit_template.cpp [hunk 1]
```diff
@@ -11,8 +11,13 @@ sections with user provided statements.
 #include <pybind11/pybind11.h>
 #include <torch/extension.h>
 
+// Global setup. (e.g. #includes)
+// GLOBAL_SETUP_TEMPLATE_LOCATION
+
 
 double timeit(int n) {
+    pybind11::gil_scoped_release no_gil;
```
Comment (swolchok) [inline torch/utils/benchmark/utils/timeit_template.cpp:19]: this change doesn't seem related to the rest?

Comment (robieta) [reply torch/utils/benchmark/utils/timeit_template.cpp]: Yeah, but I didn't really have a good place to put it. I've moved it to https://github.com/pytorch/pytorch/pull/51121 which should be more appropriate.

File: torch/utils/benchmark/utils/cpp_jit.py [hunk 2]
```diff
@@ -98,8 +98,15 @@ def get_compat_bindings() -> CallgrindModuleType:
     return COMPAT_CALLGRIND_BINDINGS
 
 
-def _compile_template(stmt: str, setup: str, src: str, is_standalone: bool) -> Any:
+def _compile_template(
```
Comment (ezyang) [inline torch/utils/benchmark/utils/cpp_jit.py:101]: You should probably just make this kwarg only, serious str blindness here

File: torch/utils/benchmark/utils/timer.py [hunk 3]
```diff
@@ -187,16 +195,21 @@ def __init__(
...
         # (For instance, `eval` adds the `__builtins__` key)
         self._globals = dict(globals or {})
+
+        timer_kwargs = {}
         if language in (Language.PYTHON, "py", "python"):
             # Include `torch` if not specified as a convenience feature.
             self._globals.setdefault("torch", torch)
             self._language: Language = Language.PYTHON
+            if global_setup is not None:
+                raise ValueError(f"global_setup is C++ only, got `{global_setup}`")
```
Comment (ezyang) [inline torch/utils/benchmark/utils/timer.py:205]: Wouldn't `global_setup` work just fine in Python too, though?

**Review by ezyang — APPROVED** [review]
I didn't check to see if there were any places where you had accidentally transposed two string arguments; typechecker won't help you here. There aren't any tests and so it is also hard to tell how you know the new logic is working or not.
